File: backend/src/domain/entities/workflow.py
import uuid
from datetime import datetime
from typing import Any, Dict, Optional

from pydantic import BaseModel, ConfigDict, Field


class WorkflowBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    definition: Dict[str, Any] = Field(default_factory=dict)


class WorkflowCreate(WorkflowBase):
    pass


class WorkflowUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    definition: Optional[Dict[str, Any]] = None
    state: Optional[str] = Field(None, pattern="^(draft|active|disabled)$")


class WorkflowResponse(WorkflowBase):
    id: uuid.UUID
    owner_id: Optional[uuid.UUID] = None
    state: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScanRunResponse(BaseModel):
    id: uuid.UUID
    workflow_id: Optional[uuid.UUID] = None
    scope_id: Optional[uuid.UUID] = None
    plugin_id: Optional[uuid.UUID] = None
    type: str
    status: str
    start_ts: Optional[datetime] = None
    end_ts: Optional[datetime] = None
    metrics: Optional[Dict[str, Any]] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkflowEventResponse(BaseModel):
    id: uuid.UUID
    workflow_id: uuid.UUID
    event_type: str
    correlation_id: Optional[uuid.UUID] = None
    payload: Optional[Dict[str, Any]] = None
    timestamp: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkflowStartRequest(BaseModel):
    scope_id: uuid.UUID
    options: Optional[Dict[str, Any]] = Field(default_factory=dict)


class WorkflowStartResponse(BaseModel):
    workflow_id: uuid.UUID
    run_id: uuid.UUID
    status: str

File: frontend/src/types/workflow.ts
export interface Workflow {
  id: string;
  name: string;
  definition: Record<string, any>;
  owner_id?: string | null;
  state: 'draft' | 'active' | 'disabled';
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  definition: Record<string, any>;
}

export interface WorkflowUpdate {
  name?: string;
  definition?: Record<string, any>;
  state?: 'draft' | 'active' | 'disabled';
}

export interface ScanRun {
  id: string;
  workflow_id?: string | null;
  scope_id?: string | null;
  plugin_id?: string | null;
  type: string;
  status: string;
  start_ts?: string | null;
  end_ts?: string | null;
  metrics?: Record<string, any> | null;
  created_at: string;
}

export interface WorkflowEvent {
  id: string;
  workflow_id: string;
  event_type: string;
  correlation_id?: string | null;
  payload?: Record<string, any> | null;
  timestamp: string;
}

export interface WorkflowStartRequest {
  scope_id: string;
  options?: Record<string, any>;
}

export interface WorkflowStartResponse {
  workflow_id: string;
  run_id: string;
  status: string;
}
